### DISCUSSION TITLE
[Idea] Broker Expression2 to be a separate new language on Github linguist

### DISCUSSION THREAD

**Discussion by TAbdiukov** [wiremod/wire#2877] [Suggestions]
I use LUA as a 'language' for my E2 code, which isn't really correct. I suggest making a Wiremod-wide effort to broker Expression2 to be a new separate language recognized by Github

[github-linguist](https://github.com/github-linguist)

**Comment by vurvdev** [comment]
This isn't gonna happen for a few big reasons, and I did want to do this years ago before I started making all these changes to wiremod:

1. E2 files are stored in .txt, so it may fail the criteria of linguist of being identifiable. I doubt they want to add a game scripting language to .txt. Even if this wasn't an issue of the extension, it makes it significantly harder to identify just how many e2 chips are out there to fit their other criteria of X amount of results on github.

2. E2's syntax is changing heavily at the moment, so doing this would actually be horrible timing. (Lambdas added, events semi-recently added, `let` and `const` added)

3. To follow 2, the syntax is being made closer to typescript, with one reason being for better syntax highlighting. So there goes the biggest reason to try and push for e2 on linguist.

I'd advise you just use typescript instead

**Reply by Denneisk** [reply]
Too bad you can't use arbitrary file extensions on Gmod (it should be a blacklist, not a whitelist imo).

**Reply by vurvdev** [reply]
> Too bad you can't use arbitrary file extensions on Gmod (it should be a blacklist, not a whitelist imo).

Yeah but facepunch devs are too busy pretending they care about security to do that

**Comment by TAbdiukov** [comment]
> 1. E2 files are stored in .txt, so it may fail the criteria of linguist of being identifiable

As a compensating control, I found the following `.gitattributes` line to be effective,
```
*2/*.txt linguist-language=Lua
```
(Lua/Typescript is Expression2 'alternative' here)

> 2. E2's syntax is changing heavily at the moment, so doing this would actually be horrible timing. (Lambdas added, events semi-recently added, let and const added)

Fair, however, perhaps these actions are asynchronous from one another? I.e. perhaps Expression2 recognition by github-linguist comes separately from language structure changes?

**Reply by vurvdev** [reply]
> As a compensating control, I found the following .gitattributes line to be effective,

You can just use this
```gitattributes
*.txt linguist-language=golo
*.txt golo
```
Also curious why you choose lua when `golo` or now `typescript` fit much better

> Fair, however, perhaps these actions are asynchronous from one another? I.e. perhaps Expression2 recognition by github-linguist comes separately from language structure changes?

They are not separate at all, linguist is strongly tied to the syntax of a language, you need to provide a highlighter for the language, and for languages with extension conflicts, a regex to identify the language.

I don't think it'd be a good idea to merge the current syntax when still more changes are yet to come, can think of a lot of things that could come in the near future that'd change the syntax (`yield`, `server` / `client`, decorators, format strings)

Also this is completely ignoring whether they'd merge it once again, which I say is very unlikely

**Reply by TAbdiukov** [reply]
I'm afraid you missed the point, see

> **`*2/*.txt `**linguist-language=...

This `*2/*.txt` can be an imperfect but practical way to detect Expression2 code

As for LUA vs Golo vs Typescript - I opted for LUA detection, because Expression2 code ultimately runs upon LUA. But yes, good arguments can be made (e.g. Lambda, local variables, etc) for TS/Golo. But,  I actually think (and hence the idea( that Expression2 should be its own little language recognized by [github-linguist](https://github.com/github-linguist)
